src/script/lua_api/l_object.cpp

```diff
@@ -1806,6 +1806,10 @@ int ObjectRef::l_set_sky(lua_State *L)
 		lua_getfield(L, 2, "fog");
 		if (lua_istable(L, -1)) {
 			sky_params.fog_distance = getintfield_default(L, -1,  "fog_distance", sky_params.fog_distance);
+			if (sky_params.fog_distance >= 0) {
+				// if set, enforce game limits
+				sky_params.fog_distance = rangelim(sky_params.fog_distance, 20, 4000);
+			}
```

Comment:
I think the minimum is too high. A fog distance of 10 allows for really nice effects, but there could also be use cases for something like 2. The only value I really see no use for is 0.

I don't think we should define an artificial minimum here. Instead, we could just change this

> Any value **>= 0** sets the desired upper bound for the client's viewing_range and disables range_all.

into this

> Any value **> 0** sets the desired upper bound for the client's viewing_range and disables range_all.

and call it day. (Not only in the docs, but also in the implementation, of course.)

Reply:
The client won't allow less than 20.

https://github.com/minetest/minetest/blob/25ef8f39341724550c3af19873e7f324a5a57251/src/client/game.cpp#L2513

Reply:
The client won't allow the user to set a value below 20, but via the current `set_sky` API, you can set values below 20 and they will be applied. E.g. this works:

```
minetest.register_on_joinplayer(function(player)
    player:set_sky({ fog = {fog_distance = 2 }})
end)
```

And I think there's no good reason to disallow it.

Reply:
I do not feel strongly. The main part to avoid was allowing setting this to 0 - which would not render anything.

Reply:
I'd like to get some consensus. I can change this to only disallow 0 as valid value.

Reply:
Just to show a possible use case for lower values: I could imagine `player:set_sky({ fog = { fog_distance = 5, fog_start = 0 }, type = "plain", base_color = "#4a4a44", clouds = false })` as an alternative implementation of smoke grenades in the "Capture the Flag" game.

![screenshot](https://user-images.githubusercontent.com/82708541/250632347-78f20fb3-e08c-45bf-906d-60d8702ec92e.png)

Btw, I really appreciate this feature.

Reply:
I don't see an issue with setting `fog_distance` even to 0. Why do we need a limit?

Reply:
At 0 you see nothing. Even the wield item is not shaded correctly.
It's not just black (or fog color) it looks like a bug.

Reply:
> At 0 you see nothing. Even the wield item is not shaded correctly.
>  It's not just black (or fog color) it looks like a bug.

Some more testing reveals that for me, any `fog_distance` value below 1 results in only the fog color being visible, both with shaders and with the fixed pipeline. The value of `fog_start` doesn't seem to make any difference.

```lua
-- for Minetest game
minetest.register_on_joinplayer(function(player)
    minetest.after(0, function()
        player:set_sky({ fog = { fog_distance = <insert value here>, fog_start = 0.5 }, type = "plain", base_color = "#ff0000", clouds = false })
    end)
end)
```

All screenshots were taken from the same point of view. The first screenshot is always with shaders (`enable_shaders = true`) and the second with the fixed pipeline (`enable_shaders = false`). The differences are probably (partially) due to the fact that tone mapping is enabled (`tone_mapping = true`).

`fog_distance = 10`
<img src="https://github.com/minetest/minetest/assets/82708541/e403ae5b-7227-497a-a033-3270f1337a7b" alt="screenshot with fog_distance set to 10, shaders enabled" height="256" />
<img src="https://github.com/minetest/minetest/assets/82708541/57e983b3-cf1d-46d0-9b18-992991317ce8" alt="screenshot with fog_distance set to 10, shaders disabled" height="256" />

`fog_distance = 1`
<img src="https://github.com/minetest/minetest/assets/82708541/68b38f57-49bf-40b6-8fff-b40e3d9bf5b3" alt="screenshot with fog_distance set to 1, shaders enabled" height="256" />
<img src="https://github.com/minetest/minetest/assets/82708541/a6d66769-8317-4bd4-89fb-f57e62b5b6cc" alt="screenshot with fog_distance set to 1, shaders disabled" height="256" />

`fog_distance = 0.999999`
<img src="https://github.com/minetest/minetest/assets/82708541/53b1965a-36eb-4987-bb5b-32bc2c20275d" alt="screenshot with fog_distance set to 0.999999, shaders enabled" height="256" />
<img src="https://github.com/minetest/minetest/assets/82708541/54af2d3c-ee5b-4d23-aa61-d1548f031a28" alt="screenshot with fog_distance set to 0.999999, shaders disabled" height="256" />

`fog_distance = 0.1`
<img src="https://github.com/minetest/minetest/assets/82708541/8c11e6aa-2865-415b-a688-b01c1ba94688" alt="screenshot with fog_distance set to 0.1, shaders enabled" height="256" />
<img src="https://github.com/minetest/minetest/assets/82708541/97c58639-0a83-475f-a8c0-b66e15f66530" alt="screenshot with fog_distance set to 0.1, shaders disabled" height="256" />

`fog_distance = 0`
<img src="https://github.com/minetest/minetest/assets/82708541/6178bf44-97b6-44eb-8eca-49539317aa9b" alt="screenshot with fog_distance set to 0, shaders enabled" height="256" />
<img src="https://github.com/minetest/minetest/assets/82708541/5087de2a-6da0-4936-b75f-7e80da851343" alt="screenshot with fog_distance set to 0, shaders disabled" height="256" />

So, something seems to break with values below 1.

Reply:
`fog_distance` is an s16, that's why it's rounded.

Not seeing anything bug fog is the intended behaviour for `fog_distance=0`, I'd say.
And the wieldhand issues are not that severe, imo.

Reply:
Ok... Make this simpler :)
I'll post an update soon.